batch/batch/driver/instance_monitor.py

```diff
@@ -0,0 +1,291 @@
+import re
+import asyncio
+import aiohttp
+import logging
+import collections
+import datetime
+import sortedcontainers
+import dateutil.parser
+
+from hailtop import aiotools
+from hailtop.utils import time_msecs, secret_alnum_string
+
+from .instance import Instance
+from ..batch_configuration import PROJECT
+
+log = logging.getLogger('instance_monitor')
+
+
+def parse_resource_name(resource_name):
+    match = RESOURCE_NAME_REGEX.fullmatch(resource_name)
+    assert match
+    return match.groupdict()
+
+
+RESOURCE_NAME_REGEX = re.compile('projects/(?P<project>[^/]+)/zones/(?P<zone>[^/]+)/instances/(?P<name>.+)')
+
+
+class InstanceMonitor:
+    def __init__(self, app, machine_name_prefix):
+        self.app = app
+        self.db = app['db']
+        self.compute_client = app['compute_client']
+        self.logging_client = app['logging_client']
+        self.log_store = app['log_store']
+        self.zone_monitor = app['zone_monitor']
+
+        self.machine_name_prefix = machine_name_prefix
+
+        self.instances_by_last_updated = sortedcontainers.SortedSet(
+            key=lambda instance: instance.last_updated)
+
+        self.name_instance = {}
+
+        self.n_instances_by_state = {
+            'pending': 0,
+            'active': 0,
+            'inactive': 0,
+            'deleted': 0
+        }
+
+        # pending and active
+        self.live_free_cores_mcpu = 0
+        self.live_total_cores_mcpu = 0
+
+        self.task_manager = aiotools.BackgroundTaskManager()
+
+    async def async_init(self):
+        log.info('initializing instance monitor')
+
+        async for record in self.db.select_and_fetchall(
+                'SELECT * FROM instances WHERE removed = 0;'):
+            instance = Instance.from_record(self.app, record)
+            self.add_instance(instance)
+
+        log.info('finished initializing instance monitor')
+
+    async def run(self):
```

Comment:
I had to split async_init into async_init and run because of circular dependencies between the instance monitor and the instance pool.

Reply:
I think the InstancePool (of which there are eventually many) should be a member of the InstanceMonitor.


We'll have an easier time working with this system if information flows consistently in one direction. I particularly find it confusing that instance pool's create_instance calls up to instance monitor who then calls back down to instance pool's adjust_for_add_instance.

I'm inclined for events to flow down:
```
... -> InstanceMonitor -> InstancePool
```
And information to flow up:
```
InstancePool -> InstanceMonitor -> ...
```

In particular, if we assign each instance pool a unique name prefix like `001` etc., then name generation can be handled entirely in the pool. When an event comes in, the name prefix can be used to dispatch the event to the correct instance pool for handling.

The instance monitor's live_free_cores_mcpu etc. can be calculated by summing the constituent pools on demand (there shouldn't be more than 10 or so pools per monitor anyway so this isn't a huge cost).

I guess as I pull the thread on this, I don't understand the essence of InstanceMonitor. Clearly we need someone who dispatches events to all the pools. We might call that job "instance event dispatcher".

There's also this monitoring loop, which feels rather self-contained and not related to anything else. It's just a sorted set and a loop that does something like:
```python3
while True:
    if self.instances_by_updated:
        instance = self.instances_by_updated[0]
        if instance.too_old():
             instance.inst_pool.check_on(instance)  # or maybe instance.check_in()
```

What do you think of all this?